db/range_tombstone_fragmenter.cc

```diff
@@ -0,0 +1,182 @@
+//  Copyright (c) 2018-present, Facebook, Inc.  All rights reserved.
+//  This source code is licensed under both the GPLv2 (found in the
+//  COPYING file in the root directory) and Apache 2.0 License
+//  (found in the LICENSE.Apache file in the root directory).
+
+#include "db/range_tombstone_fragmenter.h"
+
+#include <set>
+#include <algorithm>
+#include <functional>
+
+#include "util/kv_map.h"
+#include "util/autovector.h"
+
+namespace rocksdb {
+
+namespace {
+
+struct ParsedInternalKeyComparator {
+  ParsedInternalKeyComparator(const InternalKeyComparator* c) : cmp(c) {}
+
+  bool operator()(const ParsedInternalKey& a,
+                  const ParsedInternalKey& b) const {
+    return cmp->Compare(a, b) < 0;
+  }
+
+  const InternalKeyComparator* cmp;
+};
+
+}
+
+FragmentedRangeTombstoneIterator::FragmentedRangeTombstoneIterator(
+    std::unique_ptr<InternalIterator> unfragmented_tombstones,
+    const InternalKeyComparator& icmp, SequenceNumber snapshot)
+    : tombstone_cmp_(icmp.user_comparator()), ucmp_(icmp.user_comparator()) {
+  if (unfragmented_tombstones == nullptr) {
+    pos_ = tombstones_.end();
+    return;
+  }
+  Slice cur_start_key(nullptr, 0);
+  auto cmp = ParsedInternalKeyComparator(&icmp);
+  std::set<ParsedInternalKey, ParsedInternalKeyComparator> cur_end_keys(cmp);
+  auto flush_current_tombstones = [&](const Slice& next_start_key) {
+    auto it = cur_end_keys.begin();
+    bool reached_next_start_key = false;
+    for (; it != cur_end_keys.end() && !reached_next_start_key; ++it) {
+      Slice cur_end_key = it->user_key;
+      if (icmp.user_comparator()->Compare(cur_start_key, cur_end_key) ==
+          0) {
+        // Empty tombstone.
+        continue;
+      }
+      if (icmp.user_comparator()->Compare(next_start_key, cur_end_key) <= 0) {
+        reached_next_start_key = true;
+        cur_end_keys.erase(cur_end_keys.begin(), it);
+        cur_end_key = next_start_key;
+      }
+      autovector<SequenceNumber> seqnums_to_flush;
+      for (auto flush_it = it; flush_it != cur_end_keys.end(); ++flush_it) {
+        seqnums_to_flush.push_back(flush_it->sequence);
+      }
+      std::sort(seqnums_to_flush.begin(), seqnums_to_flush.end(),
+                std::greater<SequenceNumber>());
+      for (auto seqnum : seqnums_to_flush) {
```

Comment:
I think I could simplify this part by only writing a single tombstone for this fragment using the largest seqnum.